src/Featurizers/Components/GlobalMinimumFrequencyEstimator.h

```diff
@@ -0,0 +1,200 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+
+#include "TrainingOnlyEstimatorImpl.h"
+#include "FrequencyEstimator.h"
+#include "GrainFeaturizerImpl.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+static constexpr char const * const         GlobalMinimumFrequencyEstimatorName("GlobalMinimumFrequencyEstimator");
+static constexpr char const * const         GrainedFrequencyEstimatorName("GrainedFrequencyEstimator");
+
+template <size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()>
+using GrainedFrequencyEstimator = GrainEstimatorImpl<std::vector<std::string>, FrequencyEstimator<>>;
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         GlobalMinimumFrequencyAnnotation
+///  \brief         This is an annotation class which holds the frequency
+///                 of TimeSeries
+///
+class GlobalMinimumFrequencyAnnotation{
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using FrequencyType                     = std::chrono::system_clock::duration;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    FrequencyType                           Value;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    GlobalMinimumFrequencyAnnotation(FrequencyType value);
+    ~GlobalMinimumFrequencyAnnotation(void) = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(GlobalMinimumFrequencyAnnotation);
+};
+
+namespace Details {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         GlobalMinimumFrequencyTrainingOnlyPolicy
+///  \brief         `GlobalMinimumFrequencyEstimator` implementation details.
+///
+template <typename EstimatorT>
+class GlobalMinimumFrequencyTrainingOnlyPolicy {
+public:
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                     = std::tuple<std::vector<std::string>, std::chrono::system_clock::time_point>;

```

Comment:
In my ROllingWIndow I am defining the time_point as the first input in the tuple. Do you remember what you are doing in your laglead? we need to be consistent. Also, there will be other inputs in the featurizers that follow this. Do we need to account for them here?

Reply:
Just wondering, if you declare it is a tuple with a vector, then you need to copy all that data to populate the vector to put it in the tuple. If you continue down this road, neither ORT nor C++ will make ML.NET algorithms fast. Can you make input a structure with iterators that point to the data? We already did that.

---
In reply to: [389184166](https://github.com/microsoft/FeaturizersLibrary/pull/160#discussion_r389184166) [](ancestors = 389184166)

Reply:
Using `std::vector<std::string>` to represent a grain is a hack introduced with TimeSeriesImputer that is now breaking down. We need to address this before the next release, but is something that we can move forward with now knowing that it has to be addressed in the near future.

---
In reply to: [389185922](https://github.com/microsoft/FeaturizersLibrary/pull/160#discussion_r389185922) [](ancestors = 389185922,389184166)